### ISSUE TITLE
Train over multi-class classification - Error

### ISSUE THREAD

**Issue by samanemami** [zzd1992/GBDTMO-EX#6]
I tried to fit the model over a multi-class synthetic dataset as this [gist](https://gist.github.com/samanemami/ae9b433e2f1bceaed71f430a1fbdcd02).
But every time, it gives me the following error;
```
Error: Session cannot generate requestsError: Session cannot generate requestsat w.executeCodeCell 
(/home/oem/.vscode/extensions/ms-toolsai.jupyter-2021.8.2041215044/out/client/extension.js:52:301310)at w.execute 
(/home/oem/.vscode/extensions/ms-toolsai.jupyter-2021.8.2041215044/out/client/extension.js:52:300703)at w.start 
(/home/oem/.vscode/extensions/ms-toolsai.jupyter-2021.8.2041215044/out/client/extension.js:52:296367)at runMicrotasks 
(<anonymous>)at processTicksAndRejections (internal/process/task_queues.js:93:5)at async 
t.CellExecutionQueue.executeQueuedCells (/home/oem/.vscode/extensions/ms-toolsai.jupyter-
2021.8.2041215044/out/client/extension.js:52:311160)at async t.CellExecutionQueue.start (/home/oem/.vscode/extensions/ms-
toolsai.jupyter-2021.8.2041215044/out/client/extension.js:52:310700)
```

I tried different datasets (Classification) and different hyperparameters.
I am using `ce` as the loss function which I assumed is the cross-entropy and I think this causes the error.

**Comment by zzd1992** [comment]
You should set `out_dim` correctly.
`out_dim` should be the number of classes.
In your case, `out_dim=0`. I think this is the reason.

**Comment by samanemami** [comment]
Thanks for the reply and your support.
I updated the out_dim to the number of classes, but it still gives me the same error.
I only have this error for classification, and for regression, it works fine.

Regarding the `loss function`, where can I find the available `string` options (`mse`, `ce`, ...)?

**Comment by zzd1992** [comment]
Please refer the `Objective` struct in `src/loss.h`.
The type of loss function is quite limited: `ce bce mse`.

But you can custom your loss in python. See `docs` for detail.

**Comment by samanemami** [comment]
Thanks. I found the `src/loss.h` a complete structure.
The thing is, the error is still here with the fact that I used all appropriate loss functions for the classification.

**Comment by zzd1992** [comment]
Could you show me the python traceback?

**Comment by samanemami** [comment]
Please find the updated gist with trackback as the [link](https://gist.github.com/samanemami/ae9b433e2f1bceaed71f430a1fbdcd02).

**Comment by zzd1992** [comment]
Looks like only the information of ipykernel.
Have you tried random dataset which I provide an example previously?
If the random dataset works well, I suggust you check whether x is contiguous after split?

**Comment by zzd1992** [comment]
Looks like only the information of ipykernel.
Have you tried random dataset which I provide an example previously?
If the random dataset works well, I suggust you check whether x is contiguous after split?

**Comment by samanemami** [comment]
@zzd1992 Thank you for review it.
The dataset is a random synthetic classification, also I used the `mnist` from Keras and returns the same error.
There is no specific error and does not work.

**Comment by samanemami** [comment]
I have done the same experiment for the iris problem, and the model again does not work for this classification problem.
It seems that the problem comes from the loss function, which I do not have access to.

**Comment by zzd1992** [comment]
I think we can't locate the problem because the traceback has no information about the program itself.
I suggest you run the script in your terminal to see which line causes the error.

You can first run this script:
```python
from gbdtmo import GBDTMulti, load_lib
import numpy as np

LIB = load_lib("path of gbdtmo.so")
inp_dim, out_dim = 10, 5
params = {"max_depth": 5, "lr": 0.1, 'loss': b"mse"}
booster = GBDTMulti(LIB, out_dim=out_dim, params=params)

x_train, y_train = np.random.rand(10000, inp_dim), np.random.rand(10000, out_dim)
x_valid, y_valid = np.random.rand(10000, inp_dim), np.random.rand(10000, out_dim)
booster.set_data((x_train, y_train), (x_valid, y_valid))
booster.train(10)
```
If this works well, GBDT-MO is installed correctly.
Then you can run your own script.
I guess the problem is caused by train_test_split function. You should check whether x is contiguous after the split.

**Comment by zzd1992** [comment]
I think we can't locate the problem because the traceback has no information about the program itself.
I suggest you run the script in your terminal to see which line causes the error.

You can first run this script:
```python
from gbdtmo import GBDTMulti, load_lib
import numpy as np

LIB = load_lib("path of gbdtmo.so")
inp_dim, out_dim = 10, 5
params = {"max_depth": 5, "lr": 0.1, 'loss': b"mse"}
booster = GBDTMulti(LIB, out_dim=out_dim, params=params)

x_train, y_train = np.random.rand(10000, inp_dim), np.random.rand(10000, out_dim)
x_valid, y_valid = np.random.rand(10000, inp_dim), np.random.rand(10000, out_dim)
booster.set_data((x_train, y_train), (x_valid, y_valid))
booster.train(10)
```
If this works well, GBDT-MO is installed correctly.
Then you can run your own script.
I guess the problem is caused by train_test_split function. You should check whether x is contiguous after split.

**Comment by samanemami** [comment]
> I think we can't locate the problem because the traceback has no information about the program itself.
> I suggest you run the script in your terminal to see which line causes the error.
> 
> You can first run this script:
> 
> ```python
> from gbdtmo import GBDTMulti, load_lib
> import numpy as np
> 
> LIB = load_lib("path of gbdtmo.so")
> inp_dim, out_dim = 10, 5
> params = {"max_depth": 5, "lr": 0.1, 'loss': b"mse"}
> booster = GBDTMulti(LIB, out_dim=out_dim, params=params)
> 
> x_train, y_train = np.random.rand(10000, inp_dim), np.random.rand(10000, out_dim)
> x_valid, y_valid = np.random.rand(10000, inp_dim), np.random.rand(10000, out_dim)
> booster.set_data((x_train, y_train), (x_valid, y_valid))
> booster.train(10)
> ```
> 
> If this works well, GBDT-MO is installed correctly.
> Then you can run your own script.
> I guess the problem is caused by train_test_split function. You should check whether x is contiguous after split.

I checked the input after the split and it is contiguous.

If I understood correctly, GBDTMO using the XGBoost formulation for the loss function, so it should not be any different in terms of using the loss, but the problem comes from the loss function.

Of course, I tried your code, and it shows that the GBDTMO has been installed correctly as I already trained the regression problem.

The problem I have is with the classification, even with the simple ones.

Have you trained the model over the iris problem?
For me, it does not work.

**Comment by zzd1992** [comment]
The problem is that:
when you run classifications with ce loss. y should be an interger array which means the index of the correct class. In your gist, y is a float array.

**Comment by samanemami** [comment]
I'm afraid that I change the y.astype to int32, and the problem is still there.
```Python
X, y = X.astype('float64'), y.astype('int32')
```

**Comment by samanemami** [comment]
It worked, thanks @zzd1992
I had to reload my editor.
This is what I used in case someone has the same issue;

```Python
X, y = X.astype('float64'), y.astype('int32')
```
One other question, the model returned the predicted value for each class, so can we conclude you label binarized the class column?
I want to reverse it to have an array with 1 row with the size of `y`. Would you please help with how you converted the `y` in classification? As I will know how can reverse it.
I appreciate it

**Comment by zzd1992** [comment]
Let us consider a 10 way classification. Suppose we have n samples. Then the input y is an array of size n. Each element of y ranges from 0 to 9. I think this is the same as Xgboost.

For prediction, it is a float array with n rows and 10 columns. Each row is the logits of 10 classes.

**Comment by samanemami** [comment]
Thank you so much
